### DISCUSSION TITLE
Writing a custom loss function

### DISCUSSION THREAD

**Discussion by Lukasnol** [MilesCranmer/PySR#906] [Q&A]
Hello there,

i am currently trying to write a custom loss function. The function is supposed to calculate some landscape features of the current term of the tree and compare it with some target features. The distance is the loss of the function. The dimensionality is 2D

I am struggeling to implement this in Julia since I am new to that language. Maybe someone can give me a helping hand. 
To calculate the landscape features i need to use the pflacco package.
It works like the following example:

```
import pflacco as pf
ela_level = pf.calculate_ela_level(X,y,function)
#X,y can be list or pandas.DataFrame or numpy array, function is the term of the tree
# the function should take [x1,x2] as an input and return y
ela_loss = (ela_level - ela_level_target)^2

return ela_loss

```
This is only an example since i want to calculate multiple features. They are all calculated by calling a method from the pflacco package, which always works in a similar way. 

In the documentation i found the way of creating a custom loss function by using julia snippets. 
I am unsure though how i can:
1. accessthe target feature vector in the function
2. use the python pflacco package inside the julia code
3. pass the tree term as a function to the pf.calculate_ela methods in a way that it takes a vector of x1,x2 as an input and generates y as an output. 

I would greatly appreciate your help. 

Thank you, 
Lukas

**Comment by MilesCranmer** [comment]
Is there a Julia version of flacco? Does anything in https://github.com/JuliaNLSolvers/Optim.jl or https://github.com/SciML/Optimization.jl do the same thing?

Generally you should avoid making calls to Python inside the loss function, because the Python GIL https://wiki.python.org/moin/GlobalInterpreterLock fundamentally prevents parallelism. It's better to find a Julia equivalent as then the search can still run in parallel.

You can also try to write it in Python and then ask ChatGPT to translate it to Julia.

**Reply by Lukasnol** [reply]
Thank you for your response. For some testing i switched to DEAP
